File: elite/state.py
"""Thread-safe store for everything the app knows about the game."""

import threading
from collections import deque


class AppState:
    def __init__(self):
        self._lock = threading.Lock()

        # Commander / ship
        self.commander = None
        self.ship_type = None
        self.ship_name = None
        self.ship_ident = None
        self.cargo_capacity = None
        self.max_jump_range = None
        self.fuel_capacity = None
        self.rebuy = None  # insurance cost from Loadout

        # Location
        self.system = None
        self.system_address = None
        self.star_pos = None
        self.body = None
        self.docked = False
        self.station = None
        self.station_type = None
        self.station_market_id = None
        self.dist_from_star_ls = None

        # Live status (Status.json)
        self.credits = None
        self.fuel_main = None
        self.fuel_reservoir = None
        self.cargo_tons = None
        self.legal_state = None
        self.destination = None

        # Collections
        self.jump_history = deque(maxlen=20)  # newest first
        self.cargo_inventory = []
        self.market = None  # {"market_id", "station", "system", "timestamp", "items": [...]}

        # Flight safety: plotted route (for fuel-scoop callouts) and a queue of
        # one-shot voice alerts (interdiction, hull damage, first discovery).
        self.nav_route = []            # [{"system","address","star_class"}, ...]
        self.fuel_used_samples = deque(maxlen=5)  # recent FSDJump FuelUsed values
        self.alerts = deque(maxlen=8)  # [{"id","level","code","say","text","ts"}]
        self._alert_seq = 0

        # Exobiology (current system signals; vault persists until sold/death)
        self.bio_signals = {}   # body name -> {count, genuses:[...], body details}
        self.bio_sampling = None  # {"genus","species","variant","progress"}
        self.bio_vault = []     # [{"species","genus","variant","value","body"}]
        # Community-mapped genuses for the current system (Spansh galaxy dump)
        self.bio_community = {}  # {"id64", "system", "bodies": {body: {count, genuses}}}

        # Colonization construction depots (latest event per MarketID)
        self.colonisation = {}  # market_id -> {progress, resources, station, ...}

        # Unsold cartographic data (cleared on sell/death)
        self.explo_scans = {}  # body name -> {base, first, mapped, class}

        # Active missions (MissionID -> details) and engineering materials
        self.missions = {}
        self.materials = {"Raw": {}, "Manufactured": {}, "Encoded": {}}

        # Combat: session counters + kills per faction while that faction has
        # active massacre missions (drives the stack-progress card).
        self.combat_kills = 0
        self.combat_bounty_cr = 0
        self.combat_bonds_cr = 0
        self.faction_kills = {}  # target faction -> kills while stack active

        # Live session counters (reset each LoadGame / game launch)
        self.session_start_ts = None      # epoch when the session began
        self.session_start_credits = None  # balance at session start
        self.session_jumps = 0
        self.session_ly = 0.0

        self.last_journal_event = None  # timestamp string of most recent event seen
        self.journal_dir_found = True

    def update(self, **kwargs):
        with self._lock:
            for key, value in kwargs.items():
                setattr(self, key, value)

    def add_jump(self, system, dist, timestamp):
        with self._lock:
            self.jump_history.appendleft(
                {"system": system, "dist": dist, "timestamp": timestamp}
            )
            self.session_jumps += 1
            if dist:
                self.session_ly += dist

    def record_kill(self, victim_faction, bounty_cr=0, bond_cr=0, counts_for_stack=False):
        """Session combat counters; returns the faction's new stack-kill count
        when the kill counts toward an active massacre stack, else None."""
        with self._lock:
            self.combat_kills += 1
            self.combat_bounty_cr += bounty_cr
            self.combat_bonds_cr += bond_cr
            if counts_for_stack and victim_faction:
                self.faction_kills[victim_faction] = self.faction_kills.get(victim_faction, 0) + 1
                return self.faction_kills[victim_faction]
        return None

    def push_alert(self, level, code, say, text):
        """Queue a one-shot voice alert. The UI speaks any alert whose id is
        newer than the last it announced, so each fires exactly once."""
        with self._lock:
            self._alert_seq += 1
            self.alerts.append({
                "id": self._alert_seq, "level": level, "code": code,
                "say": say, "text": text,
            })

    def add_fuel_used(self, tons):
        with self._lock:
            if tons and tons > 0:
                self.fuel_used_samples.append(float(tons))

    def _fuel_per_jump(self):
        # Worst of the recent jumps → a conservative "jumps of fuel" projection.
        return max(self.fuel_used_samples) if self.fuel_used_samples else None

    def _nav_snapshot(self):
        from . import flight

        ahead = flight.route_ahead(self.nav_route, self.system_address, self.system)
        fpj = self._fuel_per_jump()
        jumps_of_fuel = (
            int(self.fuel_main / fpj) if (fpj and self.fuel_main is not None) else None
        )
        # No fuel nagging while docked — you're safe and can refuel.
        advisory = (
            None if self.docked
            else flight.fuel_advisory(ahead, self.fuel_main, self.fuel_capacity, fpj)
        )
        return {
            "system": self.system,
            "ahead": ahead[:12],
            "fuel_per_jump": round(fpj, 2) if fpj else None,
            "jumps_of_fuel": jumps_of_fuel,
            "advisory": advisory,
        }

    def start_session(self, ts, credits):
        """Reset the live-session counters at a game launch (LoadGame)."""
        with self._lock:
            self.session_start_ts = ts
            self.session_start_credits = credits
            self.session_jumps = 0
            self.session_ly = 0.0
            self.combat_kills = 0
            self.combat_bounty_cr = 0
            self.combat_bonds_cr = 0

    def _massacre_snapshot(self):
        """Stack progress per target faction. Kills count toward every giver's
        massacre missions simultaneously, so the kills actually needed are the
        *largest single giver's* total, not the sum across givers."""
        stacks = {}
        for m in self.missions.values():
            if m.get("kind") != "combat" or not m.get("target_faction") or not m.get("kill_count"):
                continue
            if "massacre" not in (m.get("name") or "").lower():
                continue
            s = stacks.setdefault(m["target_faction"], {"missions": 0, "reward": 0, "by_giver": {}})
            s["missions"] += 1
            s["reward"] += m.get("reward") or 0
            giver = m.get("faction") or "?"
            s["by_giver"][giver] = s["by_giver"].get(giver, 0) + m["kill_count"]
        out = []
        for faction, s in stacks.items():
            needed = max(s["by_giver"].values())
            done = self.faction_kills.get(faction, 0)
            out.append({
                "faction": faction,
                "missions": s["missions"],
                "givers": len(s["by_giver"]),
                "reward": s["reward"],
                "kills_needed": needed,
                "kills_done": min(done, needed),
                "complete": done >= needed,
            })
        return sorted(out, key=lambda s: -s["reward"])

    def _combat_snapshot(self):
        return {
            "kills": self.combat_kills,
            "bounty_cr": self.combat_bounty_cr,
            "bonds_cr": self.combat_bonds_cr,
            "massacre": self._massacre_snapshot(),
        }

    def _session_snapshot(self):
        credits_now = self.credits
        earned = (
            credits_now - self.session_start_credits
            if credits_now is not None and self.session_start_credits is not None
            else None
        )
        return {
            "start_ts": self.session_start_ts,
            "start_credits": self.session_start_credits,
            "credits_now": credits_now,
            "earned": earned,
            "jumps": self.session_jumps,
            "ly": round(self.session_ly, 1),
        }

    def _exploration_snapshot(self):
        from . import exploration

        entries = [
            {**e, "value": exploration.effective_value(e)} for e in self.explo_scans.values()
        ]
        entries.sort(key=lambda e: -e["value"])
        return {
            "total": sum(e["value"] for e in entries),
            "count": len(entries),
            "mapped": sum(1 for e in entries if e.get("mapped")),
            "firsts": sum(1 for e in entries if e.get("first")),
            "top": entries[:8],
        }

    def snapshot(self):
        with self._lock:
            market = None
            if self.market:
                market = dict(self.market)
                market["is_current_station"] = (
                    self.docked and self.market.get("market_id") == self.station_market_id
                )
            return {
                "commander": self.commander,
                "ship_type": self.ship_type,
                "ship_name": self.ship_name,
                "ship_ident": self.ship_ident,
                "cargo_capacity": self.cargo_capacity,
                "max_jump_range": self.max_jump_range,
                "fuel_capacity": self.fuel_capacity,
                "rebuy": self.rebuy,
                "system": self.system,
                "star_pos": self.star_pos,
                "body": self.body,
                "docked": self.docked,
                "station": self.station,
                "station_type": self.station_type,
                "dist_from_star_ls": self.dist_from_star_ls,
                "credits": self.credits,
                "fuel_main": self.fuel_main,
                "fuel_reservoir": self.fuel_reservoir,
                "cargo_tons": self.cargo_tons,
                "legal_state": self.legal_state,
                "destination": self.destination,
                "jump_history": list(self.jump_history),
                "cargo_inventory": list(self.cargo_inventory),
                "market": market,
                "bio": {
                    "system_signals": self._bio_signals_snapshot(),
                    "sampling": self.bio_sampling,
                    "vault": {
                        "items": list(self.bio_vault),
                        "total": sum(i.get("value") or 0 for i in self.bio_vault),
                    },
                },
                "exploration": self._exploration_snapshot(),
                "colonisation": sorted(
                    self.colonisation.values(), key=lambda c: c.get("updated") or "", reverse=True
                ),
                "missions": sorted(
                    self.missions.values(), key=lambda m: m.get("expiry_ts") or float("inf")
                ),
                "materials": self._materials_snapshot(),
                "session": self._session_snapshot(),
                "combat": self._combat_snapshot(),
                "nav": self._nav_snapshot(),
                "alerts": list(self.alerts),
                "last_journal_event": self.last_journal_event,
                "journal_dir_found": self.journal_dir_found,
            }

    def _bio_signals_snapshot(self):
        """Your own scanned signals, enriched with community-mapped genuses
        (Spansh) for bodies you haven't DSS-mapped yourself, plus community-known
        bodies you haven't even FSS'd yet."""
        community = (
            self.bio_community.get("bodies") or {}
            if self.bio_community.get("id64") == self.system_address
            else {}
        )
        merged = {}
        for name, entry in self.bio_signals.items():
            e = dict(entry)
            cg = community.get(name)
            if cg and not e.get("genuses"):
                e["community_genuses"] = cg.get("genuses") or []
            merged[name] = e
        for name, cg in community.items():
            if name in merged:
                continue
            merged[name] = {
                "body": name,
                "count": cg.get("count"),
                "genuses": [],
                "community_genuses": cg.get("genuses") or [],
                "source": "community",
            }
        return sorted(merged.values(), key=lambda b: -(b.get("count") or 0))

    def _materials_snapshot(self):
        out = {}
        total = 0
        for cat in ("Raw", "Manufactured", "Encoded"):
            items = sorted(self.materials.get(cat, {}).values(), key=lambda m: -m.get("count", 0))
            out[cat.lower()] = items
            total += sum(m.get("count", 0) for m in items)
        out["total"] = total
        return out

File: ui/src/api/contracts/state.js
/** @import {JsonObject} from "./common.js" */
/** @import {SpecialistSnapshot} from "./specialists.js" */

/**
 * @typedef {[number, number, number]} Coordinate3
 *
 * @typedef {{
 *   lat: number,
 *   lon: number,
 *   body?: string|null,
 *   radius?: number|null,
 *   heading?: number|null,
 *   altitude?: number|null,
 * }} SurfacePosition
 *
 * @typedef {{
 *   label: string,
 *   url: string,
 * }} ExternalLink
 *
 * @typedef {{
 *   system?: string|null,
 *   address?: number|null,
 *   star_class?: string|null,
 *   distance?: number|null,
 *   ts?: number|string|null,
 * }} JumpHistoryEntry
 *
 * @typedef {{
 *   symbol?: string,
 *   name?: string,
 *   count?: number,
 *   stolen?: number,
 *   mission_id?: number|null,
 * }} CargoInventoryItem
 *
 * @typedef {{
 *   id?: number|string,
 *   name?: string,
 *   symbol?: string,
 *   category?: string,
 *   buy_price?: number,
 *   sell_price?: number,
 *   demand?: number,
 *   stock?: number,
 * }} StateMarketItem
 *
 * @typedef {{
 *   market_id?: number|null,
 *   station?: string|null,
 *   system?: string|null,
 *   timestamp?: string|number|null,
 *   is_current_station?: boolean,
 *   items?: StateMarketItem[],
 * }|null} StateMarket
 *
 * @typedef {{
 *   name?: string,
 *   min_value?: number|null,
 *   max_value?: number|null,
 *   colony_m?: number|null,
 * }} BiologyGenus
 *
 * @typedef {{
 *   body?: string,
 *   count?: number,
 *   genuses?: BiologyGenus[],
 *   community_genuses?: BiologyGenus[],
 *   predicted?: BiologyGenus[],
 *   was_discovered?: boolean,
 *   landable?: boolean,
 *   source?: string,
 *   planet_class?: string,
 *   atmosphere?: string,
 *   gravity_g?: number|null,
 *   temp_k?: number|null,
 * }} BiologySignal
 *
 * @typedef {{
 *   genus?: string|null,
 *   species?: string|null,
 *   variant?: string|null,
 *   progress?: number,
 *   value?: number|null,
 *   first?: boolean,
 *   min_dist_m?: number|null,
 *   colony_m?: number|null,
 *   clear?: boolean|null,
 * }} BiologySampling
 *
 * @typedef {{
 *   species?: string,
 *   genus?: string,
 *   variant?: string|null,
 *   value?: number,
 *   body?: string|null,
 *   first?: boolean,
 * }} BiologyVaultItem
 *
 * @typedef {{
 *   system_signals: BiologySignal[],
 *   sampling: BiologySampling|null,
 *   vault: {
 *     items: BiologyVaultItem[],
 *     total: number,
 *   },
 * }} BiologyState
 *
 * @typedef {{
 *   body?: string,
 *   class?: string,
 *   mapped?: boolean,
 *   first?: boolean,
 *   value?: number,
 * }} ExplorationBody
 *
 * @typedef {{
 *   total: number,
 *   count: number,
 *   mapped: number,
 *   firsts: number,
 *   top: ExplorationBody[],
 * }} ExplorationState
 *
 * @typedef {{
 *   start_ts?: number|null,
 *   end_ts?: number|null,
 *   start_credits?: number|null,
 *   jumps?: number,
 *   distance_ly?: number,
 *   collected_cr?: number,
 * }} SessionState
 *
 * @typedef {{
 *   kills?: number,
 *   bounty_cr?: number,
 *   bonds_cr?: number,
 *   faction_kills?: {[faction: string]: number},
 * }} CombatState
 *
 * @typedef {{
 *   route?: {system?: string, address?: number|null, star_class?: string|null}[],
 *   fuel_per_jump?: number|null,
 *   scoopable_ahead?: boolean|null,
 *   next_star_class?: string|null,
 * }} NavigationState
 *
 * @typedef {{
 *   system: string|null,
 *   star_class: string|null,
 *   scoopable: boolean,
 *   taxi: boolean,
 *   started_ms: number,
 *   elapsed_s: number,
 * }} JumpProgress
 *
 * @typedef {{
 *   id?: number|string,
 *   level?: string,
 *   code?: string,
 *   say?: string,
 *   text?: string,
 *   ts?: number,
 * }} StateAlert
 *
 * @typedef {{
 *   active: boolean,
 *   phase: string,
 *   completed: number,
 *   total: number,
 *   current: string|null,
 *   attempt: number,
 *   retrying: boolean,
 * }} JournalRebuildState
 *
 * @typedef {{
 *   commander: string|null,
 *   commander_id: string|null,
 *   ship_type: string|null,
 *   ship_name: string|null,
 *   ship_ident: string|null,
 *   cargo_capacity: number|null,
 *   max_jump_range: number|null,
 *   fuel_capacity: number|null,
 *   rebuy: number|null,
 *   horizons: boolean|null,
 *   odyssey: boolean|null,
 *   game_version: string|null,
 *   game_build: string|null,
 *   has_loadout: boolean,
 *   system: string|null,
 *   system_address: number|null,
 *   star_pos: Coordinate3|null,
 *   pos: SurfacePosition|null,
 *   body: string|null,
 *   docked: boolean,
 *   station: string|null,
 *   station_type: string|null,
 *   dist_from_star_ls: number|null,
 *   credits: number|null,
 *   fuel_main: number|null,
 *   fuel_reservoir: number|null,
 *   cargo_tons: number|null,
 *   legal_state: string|null,
 *   destination: string|null,
 *   jump_history: JumpHistoryEntry[],
 *   cargo_inventory: CargoInventoryItem[],
 *   market: StateMarket,
 *   bio: BiologyState,
 *   exploration: ExplorationState,
 *   colonisation: JsonObject[],
 *   missions: JsonObject[],
 *   materials: JsonObject,
 *   synth: JsonObject,
 *   engineers: JsonObject[],
 *   galaxy: JsonObject,
 *   stored_ships: JsonObject|null,
 *   ship_locker: JsonObject|null,
 *   carrier: JsonObject|null,
 *   specialists: SpecialistSnapshot|null,
 *   session: SessionState,
 *   combat: CombatState,
 *   nav: NavigationState,
 *   jump: JumpProgress|null,
 *   alerts: StateAlert[],
 *   last_journal_event: string|null,
 *   journal_dir_found: boolean,
 *   journal_rebuild: JournalRebuildState,
 *   game_running: boolean|null,
 *   galaxy_mode: "live"|"legacy"|"unknown",
 *   links: ExternalLink[],
 * }} ApplicationState
 *
 * @typedef {{
 *   station: string,
 *   type?: string|null,
 *   body?: string|null,
 *   dist_ls?: number|null,
 *   large_pad?: boolean,
 *   pads?: {large?: number, medium?: number, small?: number},
 *   economy?: string|null,
 *   faction?: string|null,
 *   services?: string[],
 *   market_id?: number|null,
 *   local_market?: boolean,
 *   market_updated?: number|string|null,
 * }} SystemStation
 *
 * @typedef {{
 *   system: string,
 *   stations: SystemStation[],
 *   note?: string,
 * }} SystemStationsResponse
 *
 * @typedef {{
 *   ship_type: string|null,
 *   ship_name: string|null,
 *   ship_ident: string|null,
 *   edsy_url: string,
 *   slef: JsonObject,
 * }} LoadoutExportResponse
 *
 * @typedef {{
 *   ok: true,
 *   already_running?: true,
 *   via?: string,
 * }} LaunchGameResponse
 */

export {};
